### PR TITLE
Token by symbol or address

### PR DESCRIPTION
**Pull request by nlordell**
One step towards #1275 

This PR implements handling of token symbols and addresses for incoming requests. Note that there is still no implementation for mapping token symbols and addresses to token IDs, that will come in a follow up PR.

### Test Plan

```
$ cargo run -p price-estimator -- --orderbook-file target/orderbook.mainnet.bin
...
$ curl -s 'http://localhost:8080/api/v1/markets/WETH-OWL/estimated-buy-amount/1?atoms=true' | jq 
{
  "message": "token symbol or address not found"
}
```

### CODE REVIEW

File: price-estimator/src/filter.rs [hunk 1]
```diff
@@ -44,6 +48,10 @@ pub fn all(
 struct NoTokenInfo;
 impl Reject for NoTokenInfo {}
 
+#[derive(Debug)]
+struct TokenNotFound;
+impl Reject for TokenNotFound {}
```
Comment (e00E) [inline price-estimator/src/filter.rs:53]: Maybe merge it with the struct above that we already have `NoTokenInfo`. Or do we think it's important to give the user a different error message?

File: price-estimator/src/models/currency_pair.rs [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    }
+}
+
+/// A token reference, either by token ID, address, or symbol.
+#[derive(Clone, Debug, PartialEq)]
+pub enum TokenRef {
+    Id(TokenId),
+    Address(Address),
+    Symbol(String),
+}
```
Comment (e00E) [inline price-estimator/src/models/currency_pair.rs:58]: One problem I see now is that it is possible for a token symbol to also be a valid id or address. We should make sure that Id has the highest priority, then Address then Symbol. In this PR it is already correct but I'd add a comment there that the order of deserialization attempts matters.

**Review by e00E — APPROVED** [review]

File: price-estimator/src/filter.rs [hunk 3]
```diff
@@ -54,6 +62,9 @@ async fn handle_rejection(err: Rejection) -> Result<impl Reply, Infallible> {
     } else if let Some(NoTokenInfo) = err.find() {
         code = StatusCode::BAD_REQUEST;
         message = "request with atoms=true for token we don't have erc20 info for";
```
Comment (fleupold) [inline price-estimator/src/filter.rs:64]: Unrelated: It seems like this message is a) inverted and b) referencing the deprecated parameter

File: price-estimator/src/filter.rs (hunk 1 above)
Comment (fleupold) [reply price-estimator/src/filter.rs:53]: I think in one case we have found the token but don't have decimal information (all routes in atoms still work) and in the other case the token might not even be listed (no routes will work). I think it's worth while differentiating the two.

File: price-estimator/src/models/currency_pair.rs (hunk 2 above)
Comment (fleupold) [reply price-estimator/src/models/currency_pair.rs:58]: Good observation. I think this behaviour is already guaranteed/documented by the unit tests (but a comment can't hurt).

**Review by fleupold — APPROVED** [review]

File: price-estimator/src/filter.rs (hunk 1 above)
Comment (nlordell) [reply price-estimator/src/filter.rs:53]: > Maybe merge it with the struct above that we already have NoTokenInfo

I agree, we should probably only have one error type for rejections.

> I think it's worth while differentiating the two.

I think we can use an enum with different variants so we can differentiate between the errors.

Comment (e00E) [reply price-estimator/src/filter.rs:53]: These structs are only needed so we can downcast and set the correct error message. I don't think we need to create one enum that holds both NoTokenInfo and TokenNotFound but it is an idea to create a big enum that holds all errors (but not in this PR).

Comment (nlordell) [reply price-estimator/src/filter.rs:44]: I implemented this in #1299, so I'll leave the error as is for now.

File: price-estimator/src/models/currency_pair.rs (hunk 2 above)
Comment (nlordell) [reply price-estimator/src/models/currency_pair.rs:58]: Added a comment in the `FromStr` implementation.

**Pull request merged**